src/libos/src/net/socket/socket_file.rs

```diff
@@ -0,0 +1,97 @@
+use super::*;
+
+use crate::fs::{
+    occlum_ocall_ioctl, AccessMode, CreationFlags, File, FileRef, IoctlCmd, StatusFlags,
+};
+use std::any::Any;
+use std::io::{Read, Seek, SeekFrom, Write};
+
+//TODO: refactor write syscall to allow zero length with non-zero buffer
+impl File for Socket {
+    fn read(&self, buf: &mut [u8]) -> Result<usize> {
+        self.recv(buf, RecvFlags::empty())
+    }
+
+    fn write(&self, buf: &[u8]) -> Result<usize> {
+        self.send(buf, SendFlags::empty())
+    }
+
+    fn read_at(&self, offset: usize, buf: &mut [u8]) -> Result<usize> {
+        if offset != 0 {
+            return_errno!(ESPIPE, "a nonzero position is not supported");
+        }
+        self.read(buf)
+    }
+
+    fn write_at(&self, offset: usize, buf: &[u8]) -> Result<usize> {
+        if offset != 0 {
+            return_errno!(ESPIPE, "a nonzero position is not supported");
+        }
+        self.write(buf)
+    }
+
+    // TODO: use sendmsg to impl readv
+    fn readv(&self, bufs: &mut [&mut [u8]]) -> Result<usize> {
+        let mut total_len = 0;
+        for buf in bufs {
+            match self.read(buf) {
+                Ok(len) => {
+                    total_len += len;
+                }
+                Err(_) if total_len != 0 => break,
+                Err(e) => return Err(e.into()),
+            }
+        }
+        Ok(total_len)
+    }
+
+    fn writev(&self, bufs: &[&[u8]]) -> Result<usize> {
+        let mut total_len = 0;
+        for buf in bufs {
+            match self.write(buf) {
+                Ok(len) => {
+                    total_len += len;
```

Comment:
What if `len < buf.len()`? The code should be
```
Ok(len) => {
    total_len += len;
    if len < buf.len() {
        break;
    }
}
```

Reply:
I used sendmsg/recvmsg to implement readv/writev.